assembly/src/parser/token.rs

```diff
@@ -0,0 +1,811 @@
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum HexEncodedValue {
+    /// A tiny value
+    U8(u8),
+    /// A small value
+    U16(u16),
+    /// A u32 constant, typically represents a memory address
+    U32(u32),
+    /// A single field element, 8 bytes, encoded as 16 hex digits
+    Felt(Felt),
+    /// A set of 4 field elements, 32 bytes, encoded as a contiguous string of 64 hex digits
+    Word([Felt; 4]),
+}
+
+// TOKEN
+// ================================================================================================
+
+/// The token type produced by [crate::parser::Lexer], and consumed by the parser.
+#[derive(Debug, Clone)]
+pub enum Token<'input> {
+    Add,
+    Adv,
+    InsertHdword,
+    InsertHperm,
+    InsertMem,
+    AdvLoadw,
+    AdvPipe,
+    AdvPush,
+    PushExt2intt,
+    PushMapval,
+    PushMapvaln,
+    PushMtnode,
+    PushSig,
+    PushSmtpeek,
+    PushSmtset,
+    PushSmtget,
+    PushU64Div,
+    And,
+    Assert,
+    Assertz,
+    AssertEq,
+    AssertEqw,
+    Begin,
+    Caller,
+    Call,
+    Cdrop,
+    Cdropw,
+    Clk,
+    Const,
+    Cswap,
+    Cswapw,
+    Debug,
+    Div,
+    Drop,
+    Dropw,
+    Dup,
+    Dupw,
+    Dynexec,
+    Dyncall,
+    Else,
+    Emit,
+    End,
+    Eq,
+    Eqw,
+    Ext2Add,
+    Ext2Div,
+    Ext2Inv,
+    Ext2Mul,
+    Ext2Neg,
+    Ext2Sub,
+    Err,
+    Exec,
+    Export,
+    Exp,
+    ExpU,
+    FriExt2Fold4,
+    Gt,
+    Gte,
+    Hash,
+    Hperm,
+    Hmerge,
+    If,
+    ILog2,
+    Inv,
+    IsOdd,
+    Local,
+    Locaddr,
+    LocLoad,
+    LocLoadw,
+    LocStore,
+    LocStorew,
+    Lt,
+    Lte,
+    Mem,
+    MemLoad,
+    MemLoadw,
+    MemStore,
+    MemStorew,
+    MemStream,
+    Movdn,
+    Movdnw,
+    Movup,
+    Movupw,
+    MtreeGet,
+    MtreeMerge,
+    MtreeSet,
+    MtreeVerify,
+    Mul,
+    Neg,
+    Neq,
+    Not,
+    Or,
+    Padw,
+    Pow2,
+    Proc,
+    Procref,
+    Push,
+    RCombBase,
+    Repeat,
+    RpoFalcon512,
+    Sdepth,
+    Stack,
+    Sub,
+    Swap,
+    Swapw,
+    Swapdw,
+    Syscall,
+    Trace,
+    True,
+    Use,
+    U32And,
+    U32Assert,
+    U32Assert2,
+    U32Assertw,
+    U32Cast,
+    U32Div,
+    U32Divmod,
+    U32Gt,
+    U32Gte,
+    U32Lt,
+    U32Lte,
+    U32Max,
+    U32Min,
+    U32Mod,
+    U32Not,
+    U32Or,
+    U32OverflowingAdd,
+    U32OverflowingAdd3,
+    U32OverflowingMadd,
+    U32OverflowingMul,
+    U32OverflowingSub,
+    U32Popcnt,
+    U32Clz,
+    U32Ctz,
+    U32Clo,
+    U32Cto,
+    U32Rotl,
+    U32Rotr,
+    U32Shl,
+    U32Shr,
+    U32Split,
+    U32Test,
+    U32Testw,
+    U32WrappingAdd,
+    U32WrappingAdd3,
+    U32WrappingMadd,
+    U32WrappingMul,
+    U32WrappingSub,
+    U32Xor,
+    While,
+    Xor,
+    Bang,
+    ColonColon,
+    Dot,
+    Equal,
+    Lparen,
+    Minus,
+    Plus,
+    SlashSlash,
+    Slash,
+    Star,
+    Rparen,
+    Rstab,
+    DocComment(DocumentationType),
+    HexValue(HexEncodedValue),
+    Int(u64),
```

Comment:
So this enumeration represents lexical tokens, rather than semantic values, and this token in particular is used for all base-10 encoded integer literals, whether they are ultimately `Felt` or something else. The lexer doesn't restrict the value from being >= the field modulus (any value up to `u64::MAX` is valid from the lexer point of view), the parser is where those limits are enforced.

In general there is some tension between enforcing various syntax rules in the lexer, the parser, or via semantic analysis. It is generally preferable to make the lexer quite liberal in what it accepts, and do most validation during parsing/analysis, but in some cases you can produce useful errors quite early in the lexer, so you end up with some degree of validation throughout the pipeline.

In the future, whenever I get around to it, I want to refactor the parser to emit a CST (concrete syntax tree), which is then converted to the AST during semantic analysis. This will let us implement a MASM language server. In that model, we do very little validation at all during lexing and parsing of the CST, and do virtually all of it during the conversion to an AST. Or more precisely, we do some validation during parsing of the CST, but rather than emitting errors, we represent those errors in the CST itself, and perform error recovery, so that you can support things like real-time feedback in the editor as you write code (just like `rust-analyzer` does for Rust code). In any case, that's not really relevant, but wanted to mention it, because I guess the point I'm trying to make is that the lexer is _very_ agnostic about how the tokens translate into actual AST nodes, and we want to keep it that way so as to keep the lexer flexible (with little modification, it could be used to implement the new parser approach I described above, where a more strict lexer would require more significant changes).